Colours and contrast

AndreasMoesch opened the pull request:
This pull requrest replaces @jmuheim’s original pull request (#152).

Since @jmuheim isn’t active in the project anymore we decided during [hackday 9](https://github.com/Access4all/adg/wiki/Hackday-%239) that @AndreasMoesch forks this content, applies his review and then resubmits it as a new pull request.

jmuheim committed:
Add basic structure and content

jmuheim committed:
Rename folder

jmuheim committed:
Rename folder, add content

jmuheim committed:
More content

jmuheim committed:
Add more content

jmuheim committed:
Content

jmuheim committed:
Rename

jmuheim committed:
Content

jmuheim committed:
Content

jmuheim committed:
Content

jmuheim committed:
More content

jmuheim committed:
Remove tips and tricks section

AndreasMoesch committed:
Applies review by @AndreasMoesch

See @jmuheim’s original pull request: https://github.com/Access4all/adg/pull/152/files/dee4a8a832dab9d2db01df44e0f445a69fd5186b

AndreasMoesch committed:
Removes colour profile from one example image

Images don’t have an embedded colour profile per default

AndreasMoesch committed:
Replaces all ' with ’

AndreasMoesch committed:
Various corrections and edits

* Rewriting sentences
* Reordering content
* Shortening content
* Changing titles
* changing all "" into “” in the text

AndreasMoesch commented:
@backflip Sorry, I don’t know why the checks fail. The logs point to some issue with gulp. I already encountered this problem when building the site locally. I worked around it for myself by changing the gulp dependency in the package.json file (`"gulp": "^4.0.0",` instead of `"gulp": "github:gulpjs/gulp#4.0"`) but this change was never commited.

backflip commented:
@AndreasMoesch, thanks for the pointer, I have openend https://github.com/Access4all/adg/pull/189
What is the current merge convention? Do I wait for a specific person to review and approve it?

AndreasMoesch commented:
@backflip Thanks! Regarding the merge convention: We have a defined process for content-related merge requests but this doesn’t apply to bug fixes. 

So, in the absence of a defined process I’ll add my review to #189. Seeing that this is issue prevents content-related reviews (which need a the deployed version) I think this has to be merged before release 1.1.

AndreasMoesch committed:
Updates gulp dependency in package.json

This is necessary for the build process to succeed. The build process should succeed so that the changes can be reviewed.

AndreasMoesch commented:
I pushed the updated `package.json` file (same change as in #189) so that we have a preview build for this pull request that can be used for the review.

gurix commented:
Cool thanks. Do you think we could send it to Access4All for a review? If yes, drop a mail to @retoinniger.

AndreasMoesch committed:
Fixes typo

AndreasMoesch commented:
Exchange with @AndreasUe 
* Avoid duplicate content for the Colour Contrast Analyser. Mention it in this chapter but for more in-depth information about the tool link to its dedicated page under Setup/Helpter tools
* The short paragraph about colour profiles is a bit vage but leave it for now.
* Emphasize that 4.5 is the *minimal* contrast ratio for text. It’s not a target, it’s a threshold. 
* Separate the two aspects of *contrast* (colour contrast, primarily between background and foreground) and other mechanisms for differentiation, e.g. adding graphical information. From the standpoint of the WCAG these are two related but ultimately separate concerns. In this chapter, however, they get mixed up frequently (underlined links in the “requirements for text” part, the larger part of the “requirements for graphical objects”)
  * Try to merge the two “UI components” and “graphical objects” parts into a single page that focuses on the colour contrast requirements, introduced in WCAG 2.1. This is focused on contrast between foreground and background.
  * Add an additional page that essentially focuses on rule 1.4.1 of WCAG and the contrast between different sibling objects. Here we look beyond colour contrast to differentiation between graphical elements.

AndreasMoesch commented:
Additional change related to [#216: Decide how to relate to the two versions of WCAG: 2.0 and 2.1](https://github.com/Access4all/adg/issues/216): When referencing a WCAG rule that has been introduced with WCAG 2.1 (e.g. contrast for non-text elements) mention that explicitly.

AndreasMoesch committed:
Moves the CCA content to the CCA page

* Shortens the CCA paragraph in the “colours and contrast” chapter to the minimun.
* Rewrites the dedicated “Colour Contrast Analyser” page and adds the content that was removed from the “colours and contrast” chapter

AndreasMoesch committed:
Adds reminder that 3 and 4.5 are min colour contr

Adds short paragraph on the “requirements for text” page in the “colours and contrast” chapter.

AndreasMoesch committed:
Adds mentioning of v2.1 for non-text contrast

AndreasMoesch committed:
Restructures the content, adds 2 new chapters

Separates colour contrast (foreground-background) from differentiation beyond colour (between elements)
* Removes the text correction example in the text chapter because it’s not about colour contrast per se.
* Removes the current page indicator example in the UI components chapter and slightly rephrases some passages.
* Completely rewrites the graphical objects chapter
  * Removes the shop list example and pie chart example because they both blur the line between colour and non-colour contrast.
  * Adds new line chart example to graphical objects chapter
* Adds new chapter “Colour is not enough”, explaining WCAG rule 1.4.1
  * Adds new current page indicator example
  * Adds refactored pie chart example with new visuals
* Changes chapter titles overal
* Updates edit dates

AndreasMoesch committed:
Adjusts chapter path names

* Renames the chapter directories so that they names correspond more colosely to the chapter titles.
* Updates cross-links between chapters

AndreasMoesch committed:
Edits end of “how to calculate” chapter

Rephrases and mentions the chapter “Colour is not enough”

AndreasMoesch committed:
Replaces pictures in “How to calculate” chapter

Replaces the example images in the “Tone is not e key factor” section. The red in the original pictures wasn’t always the same.

AndreasMoesch committed:
Minor corrections and adjustments

AndreasMoesch commented:
@AndreasUe I have Iterated the content of this chapter according to our conversation during the hackday on April 2. Overall, the content has changed quite a bit with new text, new examples, new images, new alt text, etc.

The small stuff first
* moved most explanations about CCA to the CCA page under `Setup/Helper Tools/Colour Contrast Analyser`
* added in multiple places reminders that `4.5:1` and `3:1` are not ideal but minimal values
* Explicitly mentioned v 2.1 when referencing a WCAG rule that has been added in 2.1

Some more foundational changes were necessary to achieve a proper separation of colour contrast (mostly for background-foreground) and more complex differentiation through shapes or patterns, etc. (mostly between objects, WCAG rule 1.4.1)
* Removed the text correction example in the text chapter because it’s not about colour contrast per se, more about adding shapes. This example doesn’t show up anymore.
* Removed the current page indicator example in the UI components chapter and slightly rephrased some passages.
* Completely rewrote the graphical objects chapter
  * Removed the shop list example and pie chart example because they both blur the line between colour contrast and contrast through shape. The shop list example isn’t used anymore but the pie chart is reused in the new “Colour is not enough” chapter with some adjustments.
  * Added new example of a line chart.
* Added new chapter “Colour is not enough”, dedicated to WCAG rule 1.4.1
  * Added a new current page indicator example. 
  * Added an edited pie chart example with new visuals and partially new wording.
* Slightly changed the titles of the different chapters.
* Added some text and a link to the “Colour is not enough” chapter at the end of the first chapter about how to measure colour contrast. 

I’m handing this off to you for another review.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-calculate/README.md:
```diff
@@ -0,0 +1,57 @@
+
+![The words “Hello” in green and “Goodbye” in red written on a blue background](_media/three-colours-words.png)
+
+… others will have a very hard time differentiating between text and background. This is how it might look for someone who sees in grayscale:
```
Where does this `…` (one single character) come from? We should stick to `...` (three characters), I think.

AndreasMoesch replied:
@jmuheim This is the typographically and semantically correct way to typeset an ellipsis (on a Swiss-German macOS keyboard it’s `Alt` and `.`). IMO the question should rather be why *not* do it like that.

AndreasMoesch replied:
**To Do:** replace `…` with `...`

AndreasMoesch replied:
Replaced all occurrences of `…` with `...`. If we change this it has to be done globally.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-calculate/README.md:
```diff
@@ -0,0 +1,57 @@
+
+… others will have a very hard time differentiating between text and background. This is how it might look for someone who sees in grayscale:
+
+![The same picture with the words “Hello” and “Goodbye” but in grayscale](_media/three-colours-words-grayscale.png)
```
The same for `“` instead of `"`.

AndreasMoesch replied:
Same on my side as well 🙂. From what I’ve learned, this is the typographically correct way to typeset quotation marks in English. Doesn’t make much of a difference in alt text. I just set it like that out of habit.

AndreasMoesch replied:
**To Do:** Replace with `"` because that’s the current standard.

AndreasMoesch replied:
Replaced all occurrences o `“` and `”` with `"`. If we change this it has to be done globally.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-calculate/README.md:
```diff
@@ -0,0 +1,57 @@
+… others will have a very hard time differentiating between text and background. This is how it might look for someone who sees in grayscale:
+
+![The same picture with the words “Hello” and “Goodbye” but in grayscale](_media/three-colours-words-grayscale.png)
+
```
We could even add here another variant of the grayscale picture with some blur added, to make it even more obvious how hard for some people it is to read with low contrast.

jmuheim commented on pages/knowledge/colours-and-contrast/README.md:
```diff
@@ -0,0 +1,9 @@
+
+# Colours and contrast
+
+**Accessibility goes far beyond functionality and code. When it comes to visual appearances, colours require special consideration. The meaning of a certain colour as well as the contrast between multiple colours directly affect how well a person can perceive and understand what is on screen. Colour palettes should therefore be composed carefully and according to some basic rules.**
```
I'd love to read something like the following here, too:

> Visual content (like text, icons, form controls, charts, etc.) needs to be easily perceivable. As such, it must clearly stand out visually from its background.

This gives a very fast understanding, what typical elements we are talking about (otherwise it feels to me like a novice could think we are talking about a very abstract design concept here).

AndreasMoesch replied:
**To Do:** simplify text.

AndreasMoesch replied:
New paragraph

> Visual elements like for example text, icons, input fields and charts need to be easily perceivable. On the most basic level that means that they have to clearly stand out from their background. But it’s also important that they can be easily distinguished from similar elements around them – for example links embedded in a text paragraph. The visual contrast between different colours is a key criteria in this context. But colour contrast also has its limitations.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-calculate/README.md:
```diff
@@ -0,0 +1,57 @@
+![The same picture with the words “Hello” and “Goodbye” but in grayscale](_media/three-colours-words-grayscale.png)
+
+It’s therefore impotant to remember:
+- Always examine contrast ratios as calculated values, don’t just rely on your eyes. More on that in the next [chapter “How to exampine colour contrast”](/knowledge/colours-and-contrast/how-to-examine/).
```
Throughout the whole guide, we tried to be as "economical" with in-site links as possible, trying to **avoid** them whenever possible. Reason: the guide is laid out as a book, and the user should only in very rare situations browse back and forth within the content.

Whenever we felt it is important enough to add in-page links, we used the following two special phrases, always in brackets:

- For links pointing backwards in the content: `(if you haven't done this yet, go back and read XXX)`.
- For links pointing forward in the content: `(if you are really curious and want to learn more about this, skip forward and read XXX)`.

In the current content, I doubt that the provided in-page links are needed, as both point to pages that are very "close" (especially the first one), and still within the current chapter.

I also think that we should not provide too many "spoilers" in advance: the user hopefully reads through the current chapter as a whole, or at least gets an overview from the page titles (where "Colour is not enough" is already mentioned).

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-calculate/README.md:
```diff
@@ -0,0 +1,57 @@
+    (L1 + 0.05) / (L2 + 0.05)
+
+This results in a value ranging from `1:1` (no contrast at all) to `21:1` (the highest possible contrast). While high contrast is generally good, higher is not always better. Research indicates that a moderate contrast somewhere between the extremes may be best.
+
```
I'd love to see this sentence here again (instead of the link at the bottom of the page):

> Fortunately, we do not need to calculate anything of this manually ourselves when using an appropriate tool. Read on for more details.

AndreasMoesch replied:
**To Do:** Mention that you don’t have to calculate yourself.

AndreasMoesch replied:
Don’t

AndreasMoesch replied:
Aaah stupid… *Done* is what I meant to say

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+
+## Colour Contrast Analyser
+
+There are a lot of tools available to calculate the contrast between two colours. For Windows and macOS users we recommend the [Colour Contrast Analyser (CCA)](/setup/helper-tools/colour-contrast-analyser/).
```
In-page link: rewrite to:

`(if you haven't done this yet, go back and read [Colour Contrast Analyser](/setup/helper-tools/colour-contrast-analyser/)).`

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+## Colour Contrast Analyser
+
+There are a lot of tools available to calculate the contrast between two colours. For Windows and macOS users we recommend the [Colour Contrast Analyser (CCA)](/setup/helper-tools/colour-contrast-analyser/).
+
```
I'd love to see the screenshot of the CCA here again.

`![Colour Contrast Analyser window](_media/colour-contrast-analyser-window.png)`

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+
+The most trustworthy source is the browser’s DOM inspector which gives direct access to the values defined in the source code.
+
+Right-click on an element and choose `Inspect` (depending on your browser the term might vary) to show that element in the DOM inspector. Now you can find the element’s colour values in the `Styles` tab.
```
Instead of `depending on your browser the term might vary` I would write `the term might vary depending on your browser`.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+
+## DOM inspector
+
+The most trustworthy source is the browser’s DOM inspector which gives direct access to the values defined in the source code.
```
Please use `'` instead of `’`.

AndreasMoesch replied:
**To Do:** Apply `'` as the current standard.

AndreasMoesch replied:
Replaced all occurences of `’` with `'`. If we change this it has to be done globally.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+
+### Anti-aliasing
+
+Most modern browsers apply anti-aliasing to make the edges of rendered elements smoother (see [Anti-aliasing (Wikipedia.org)](https://en.wikipedia.org/wiki/Spatial_anti-aliasing))
```
We should reflect `Spatial` also in the link's caption.

jmuheim commented on pages/knowledge/colours-and-contrast/how-to-examine/README.md:
```diff
@@ -0,0 +1,43 @@
+
+Colour that is displayed on screen may slightly differ from its definition in the code. This is due to the various colour profiles that are used by monitors, operating systems and browsers.
+
+So if you are using colour pickers, the picked colour may differ from its true definition.
```
The following advice is a very typical one given when giving consultancy in our daily business:

> ## Conclusion	
> Working with colour pickers can save a lot of time.	
>
> In general though: if a result is very close to the required contrast limit, we recommend to re-check it by examining the real colour definitions (if available) from the DOM inspector.

I think we should keep it.

AndreasMoesch replied:
**To Do:** Add but not as “Conclusion”

AndreasMoesch replied:
Added this short paragraph

> **A general recommendation:** Use colour pickers and profit from the time they save you. But if a result is very close to the required contrast threshold, we recommend to verify it by examining the real colour definitions (if available) in the DOM inspector.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+# Colour contrast for text
+
+**There is a general minimal colour contrast level for all kinds of text. As an exception, large text can have slightly lower contrast. This requirement applies to both “real” text and rasterised text on images.**
```
Please replace `“` and `”` with `"`. ;-)

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+Large-scale text is defined as:
+
+- Either **bold** text of at least 14 points (or 18.5 pixels).
```
I'd love to keep the code examples like this:

> - For example `font-size: 14pt; font-weight: bold` (or `font-size: 18.5px; font-weight: bold`).

And

> - For example `font-size: 18pt` (or `font-size: 24px`).

It may feel obvious to professionals, but as accessibility even for long-term web designers often is a new and confusing topic at first, we still we feel that bridging the "gap" between the two worlds (accessibility and CSS) is very useful. ("Nützt's nüt, de schadt's nüt.")

AndreasMoesch replied:
**To Do:** Add this.

AndreasMoesch replied:
@jmuheim I decided to not change it after all. I think especially considering non-coders the CSS version would make things harder to understand
* It makes it really awkward to explain that these are minimum values. “At least `font-size: 14pt` or `font-size: 18.5px`” is maybe understandable but also not the most inviting sentence.
* I don’t want this to look like a code example ready for copy-pasting because people *shouldn’t* define their font-sizes in absolute values. Explaining the pros and cons of relative vs. absolute units goes beyond the scope of this topic. Writing it in text at least creates a certain layer of abstraction.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+- Either **bold** text of at least 14 points (or 18.5 pixels).
+- Or regular text of at least 18 points (or 24 pixels).
+
+(see [WCAG definition of “large scale (text)”](https://www.w3.org/TR/WCAG21/#dfn-large-scale))
```
This paragraph sort of "hangs in the air". I would write it like this:

> For more information, see [WCAG definition of "large scale (text)"](https://www.w3.org/TR/WCAG21/#dfn-large-scale).

AndreasMoesch replied:
**To Do:** Rewrite.

AndreasMoesch replied:
Done

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+(see [WCAG definition of “large scale (text)”](https://www.w3.org/TR/WCAG21/#dfn-large-scale))
+
+To illustrate this: the following image shows the full spectrum of possible contrast ratios from `1:1` on the left to `21:1` on the right.
```
As any kind of users (e.g. blind ones) must be able to read and understand the full ADG content, we need a more detailed description of the upcoming image.

Previously, it was:

> For example, the following image has a white background and shows a text with a colour gradient overlay from white to black.

I feel that even this can be improved like following:

> For example, the following image has a white background and shows the text "The quick brown fox jumps over the lazy dog". The text has a horizontal colour gradient overlay from white ("T") to black ("g").

AndreasMoesch replied:
**To Do:** Rewrite.

AndreasMoesch replied:
Rewrote it. It’s now much more verbose but should be understandable even without seeing the picture

> To illustrate this: the following image shows the  sentence "The quick brown fox jumps over the lazy dog" written on white background. It has 43 characters. The first character "T" has the colour white. The colour of following characters gets gradually darker, the last character "g" is fully black.
>
> `![Text with colour gradient overlay](_media/lazy-dog.png)`
>
> Somewhere around the "x" (the 19th character) the contrast ratio reaches `3:1`. Somewhere around the "m" (the 23rd character) we arrive at `4.5:1`. The last character has the maximum colour contrast ratio of `21:1`.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+It may not seem like much of a difference, but in fact it heavily enlargens the number of possible colour combinations.
+
+It is important to mention that the values `3:1` and `4.5:1` represent a *minimal* standard for *acceptable* contrast, they don’t represent an ideal. They should be treated as threshold and not as target; aiming higher is recommended. 
```
We seldom used `*italics*` in the ADG. I propose we use `**bold**`.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+![A word without and a word with shadow](_media/words-without-and-with-shadow.png)
+
+The shadow of the second word results in a contrast of appropriately `4.5:1`.
```
Again, I'd love to have a code example:

> In CSS, the appropriate attribute would be something like `text-shadow: 0 0 2px #000`.

AndreasMoesch replied:
**To Do:** Add example

AndreasMoesch replied:
Done.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+![White text on blue sky and white clouds](_media/beach.png)
+
+You could increase contrast by adding a text shadow as described above. Or you could add a semi-transparent darkening layer between text and background
```
Add fullstop.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## “Real” text vs. text images
+
+With “real” text, we mean text that is available as an arrangement of character codes (like this text here). It is inherently machine-readable and can be selected and copy-pasted. “Text images” on the other hand are rasterised pictures of typography, e.g. stored as JPG or PNG files. 
```
I would add "(like the example images in the text above)" at the end of the sentence.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+![White text with semi-transparent darkening background](_media/beach-with-background.png)
+
+Of course you can also combine different solutions to add up their effects.
+
```
Here I miss the sub-chapter "Adding graphical information" with info about e.g. highlighted text:

> For example, the following image shows a gray text, and some of the words are highlighted red to mark spelling mistakes...

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Minimal contrast ratio
+
+Version 2.1 of the [Web Content Accessibility Guidelines (WCAG) specify a minimal contrast ratio](https://www.w3.org/TR/WCAG21/#non-text-contrast) of `3:1` against adjacent colour(s) for all User Interface (UI) components.
```
I would use "minimal contrast ratio of `3:1` for all User Interface components" as link text.

jmuheim replied:
So far, we introduced an abbreviation once per page and then only used the abbreviation below that.

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+changed: "2019-04-21"
+---
+
+# Colour contrast for User Interface components
```
Why use caps for "U" and "I"?

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+Because of that, certain users may have a hard time finding the input fields and enter information. 
+
+They might also struggle to identify the button by its typical shape. However, the button label adds a lot of valuable information, making this case less urgent or even negligible: it has sufficient contrast, it offers a clear invitation to interact by using a verb (which is better than a generic “OK”) and its visual appearance (font, size) is different from the other text elements surrounding it ...
```
I'd love to see this added (for the button):

> the background color (that should indicate the user that the element is a button) also has a contrast ratio of only `1.5:1`

AndreasMoesch replied:
**To Do:** Ok, add to text

AndreasMoesch replied:
Done

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+Version 2.1 of the [Web Content Accessibility Guidelines (WCAG) specify a minimal contrast ratio](https://www.w3.org/TR/WCAG21/#non-text-contrast) of `3:1` against adjacent colour(s) for all User Interface (UI) components.
+
+## Example: form controls
```
I'm unsure about the wording "Example". We use "example" pretty strictly for our code examples, and we do not have any of these on this page. Could maybe "dilute" (verwässern) the strict meaning a bit?

AndreasMoesch replied:
Leave it for now

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Example: tablists
+
+Tablists convey visually that there is a list of items, and that the content for the currently active item is displayed below. The button-like appearance of the tabs indicates that a user can select them to display their associated content.
```
I'd love to add a hint about the fact that only one item can be shown at a time. Original text (can be improved):

> The button-like style of the items indicates that a user can click on them, making the clicked item the active one and displaying its content.

AndreasMoesch replied:
**To Do:** Add info that there’s only one item selectable at a time (similar to radio buttons)

AndreasMoesch replied:
New text:

> Tablists convey visually that there is a list of items, and that the content for the currently active item is displayed below. The button-like appearance of the tabs indicates that a user can select them to display their associated content. In the overall visual arrangement the active tab is put "in front" of the inactive tabs, indicating that only one tab can be selected at a time (unlike other visual patterns such as tags or labels).

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+Tablists convey visually that there is a list of items, and that the content for the currently active item is displayed below. The button-like appearance of the tabs indicates that a user can select them to display their associated content.
+
+The borders of the following tablist have insufficient contrast (`1.5:1`). Users who cannot see them (because they have a visual impairment, because their screen is exposed to direct sunlight, because the beamer has bad contrast settings, etc.) might not realise how the different pieces relate to each other and that they can interact with them.
```
Good addition with the different scenarios! 👍 But I would add "or" to the 2nd and 3rd "because".

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+![Low contrast tablist](_media/low-contrast-tablist.png)
+
+Sometimes, tablists rely on background colours instead of borders to indicate their state and functionality. Although this may look much more obvious, it doesn’t solve the problem because the contrast is still the same.
```
We should add another screenshot here with background colours that have enough contrast, e.g. where the inactive tabs have inverted colours (white text on black background).

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+And while we are at it: providing clear indicators for where an element starts and where it ends is always good style, so please also expand borders around the content area.
+
+![Low contrast tablist](_media/tablist-with-more-borders.png)
+
```
What happened to the chapter "Active indicator"? This is very important information and should not be dropped.

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Exceptions
+
+- **Text:** For text in UI components (for example labels) the same requirements apply as for text in content *unless* the whole component is disabled. (see [chapter about colour contrast for text](/knowledge/colours-and-contrast/text/))
```
I would write "like labels" instead of "for example labels".

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+- **Text:** For text in UI components (for example labels) the same requirements apply as for text in content *unless* the whole component is disabled. (see [chapter about colour contrast for text](/knowledge/colours-and-contrast/text/))
+- **Disabled components:** this could be a disabled text field or button, but also a custom control.
+  - Be careful with this one: often, the mere existence of an element conveys something to a user. For example, a form’s submit button can be disabled while the user’s input is missing (or invalid). Someone who cannot perceive the button in such a moment might draw the wrong conclusions about the form, for example assuming that it can’t be submitted at all.
```
I'm unsure whether this shortened explanation is easy enough to understand.

Original text:

> Be careful with components that have low contrast due to having "disabled" status: often, the mere existence of an element conveys something to a user. For example, if a form's submit button is disabled while the user's input is missing (or invalid), and the disabled input (both text and shape) has low contrast, some users may not see the button at all. This could lead them to skip the form, as they may think that it is not to be submitted at all.

AndreasMoesch replied:
**To Do:** Emphasize that text *and* shape could have low contrast, making the button disappear for certain people.

AndreasMoesch replied:
New text

> Be careful with this one: often, the mere existence of an element conveys something to a user. For example, a form's submit button can be disabled while the user's input is missing (or invalid). If the label and the shape of the button have insufficient contrast, some people might not be able to perceive the button at all. This can mislead them to draw the wrong conclusions about the form, for example assuming that it can't be submitted at all.

jmuheim commented on pages/knowledge/colours-and-contrast/user-interface-components/README.md:
```diff
@@ -0,0 +1,52 @@
+
+- **Text:** For text in UI components (for example labels) the same requirements apply as for text in content *unless* the whole component is disabled. (see [chapter about colour contrast for text](/knowledge/colours-and-contrast/text/))
+- **Disabled components:** this could be a disabled text field or button, but also a custom control.
+  - Be careful with this one: often, the mere existence of an element conveys something to a user. For example, a form’s submit button can be disabled while the user’s input is missing (or invalid). Someone who cannot perceive the button in such a moment might draw the wrong conclusions about the form, for example assuming that it can’t be submitted at all.
```
4 spaces to indent a list level (not 2)!

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+---
+navigation_title: "Colour contrast for text"
```
We tried to keep the navigation titles short. Adding "Colour contrast" feels redundant to me:

![image](https://user-images.githubusercontent.com/1814983/69480077-72637480-0e04-11ea-98e9-01b003d6fd4b.png)

jmuheim commented on pages/knowledge/colours-and-contrast/graphical-objects/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## Minimal contrast ratio
+
+Version 2.1 of the [Web Content Accessibility Guidelines (WCAG) specify a minimal contrast ratio](https://www.w3.org/TR/WCAG21/#non-text-contrast) of `3:1` against adjacent colour(s) for the parts, which are required to understand the meaning or content of a graphical object.
```
I would use "minimal contrast ratio of 3:1 for all User Interface components" as link text.

jmuheim replied:
We should also point out that we are referring to AA, not AAA.

AndreasMoesch replied:
**To Do:** Correct link text

AndreasMoesch replied:
done

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## Minimal contrast ratio
+
+The [Web Content Accessibility Guidelines (WCAG) specify a minimal contrast ratio](https://www.w3.org/TR/WCAG21/#contrast-minimum) of `4.5:1` against adjacent colour(s) for text in general. An exception is made for large-scale text: It is considered easier to read and can therefore be set in a lower contrast ratio of at least `3:1`.
```
I would use "minimal contrast ratio of 3:1 for all User Interface components" as link text.

AndreasMoesch replied:
**To Do:** Correct link text

AndreasMoesch replied:
done

jmuheim commented on pages/knowledge/colours-and-contrast/graphical-objects/README.md:
```diff
@@ -0,0 +1,50 @@
+
+![Line chart with one blue trend line and overall low contrast](_media/line-chart-1-low-contrast.png)
+
+The contrast ratio of the main X and Y axis is `2.8:1`, the one of blue trend line is `2.3:1`. The text labels along the X and Y axis have a contrast ratio of `3.9:1` which can also be considered too low because [text has to meet a higher contrast requirement](/knowledge/colours-and-contrast/text) of at least `4.5:1`.  
```
Why is there a hard line-break (double whitespace)?

Please remove or replace with a new paragraph.

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+
+Colour is a great tool for separating foreground from background. However, there are many cases where being visible against a background is not enough. Different elements of a user interface or a component also have to be distinguishable *from each other.* And while finding a combination of two colours with sufficient contrast is easy, finding three, four or even more colours with sufficient cont ...
+
+The [Web Content Accessibility Guidelines’s rule on the use of colour](https://www.w3.org/TR/WCAG21/#use-of-color) therefore states that colour should “not [be] used as the only visual means of conveying information, indicating an action, prompting a response, or distinguishing a visual element.”
```
Add quote paragraph (`> bla bla bla`) here.

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+
+The following picture shows a navigation bar that could be placed at the top of the page.
+
+![navigation bar with one active item, visually differentiated by its blue text colour](_media/navigation-currentItem-1.png)
```
Hm, what happened to visible alternative texts? Possibly we never had them in the first place...

We should generate `<figure>/<figcaption>` elements, if possible!

AndreasMoesch replied:
We’ll turn this into a new issue

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+![navigation bar with one active item, visually differentiated by its blue text colour](_media/navigation-currentItem-1.png)
+
+While all the labels have sufficient contrast against the white background, the colour of the active “Shop” label and the colour of the inactive labels around it only have a contrast ratio of `1.7:1`. Someone who can’t perceive colour differences very well – because of their eye sight, their screen or the surrounding light – might struggle to spot which item is currently selected.
+
```
I would add a grayscale version of the same image here.

AndreasMoesch replied:
**To Do:** Add another asset

AndreasMoesch replied:
Done

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+
+![the same navigation bar picture as before but monochrome](_media/navigation-currentItem-2.1-monochrome.png)
+
+Finally, let’s keep in mind that beyond visual appearance the differentiation also has to be encoded in the underlying markup. It should provide adequate semantic information to clearly differentiate elements also on a non-visual level. (For more information take a look at [Semantics and their importance for accessibility](/knowledge/semantics/))
```
We should add a link to visually hidden text, and maybe add a code example like this:

```html
<ul>
  <li><a href="...">New collection</a></li>
  <li>
    <a href="...">
      Shop
      <span class="visually-hidden">(active)</span>
    </a>
  </li>
  <li><a href="...">Help</a></li>
  ...
</ul>
```

AndreasMoesch replied:
**To Do:** Add link to visually hidden text section/chapter

AndreasMoesch replied:
done

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+- Orange/yellow = `1.4:1`
+- Yellow/red = `2.4:1`
+
+This is due to the fact that contrast is primarily about difference in luminance and not about difference in tone (for more information see [Calculate colour contrast](/knowledge/colours-and-contrast/how-to-calculate/)).
```
"Backwards" link: `if you haven't done this yet, go back and read`...

See #237.

jmuheim commented on pages/knowledge/colours-and-contrast/graphical-objects/README.md:
```diff
@@ -0,0 +1,50 @@
+
+- There is rarely a single variable that can be independently tweaked. Accessibility should be built into the foundations of the design.
+- There is rarely one generic solution that “just works”. The specific context and content have to be taken into account.
+- There is no strict separation between “special-purpose accessibility” and “general-purpose usability”. In practice they both influence each other.
```
We should also point out, that strictly speaking, WCAG defines alternative ways of accessing an information as enough to also meet a criterion. For example, delivering a downloadable CSV- or Excel-file with the data that underlies a chart could be seen as "alternative way". This is a dangerous one though, as not every user might be able to download and interpret such a file...

AndreasMoesch replied:
**Question:** Is this point still part of “Colours and Contrast”?

AndreasMoesch replied:
Too specific.

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+Adding distinct visual patterns can make the visual link accessible.
+
+![The same pie chart, now with dots filling one section and lines filling another](_media/pieChart-2.1.png)
+
```
I would love to have another variant here, where instead of the three circles in the legend, each element has its own shape as legend, reflected also in the pie (e.g. apple = circle, orange = square, banana = triangle).

Or we could use the apple, orange, and banana as symbols themselves.

AndreasMoesch replied:
As discussed, rather not “overload” the guide with more examples than necessary.

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+Or in charts as simple as this one it’s also possible to forego a separate legend and place the labels directly inside the sections. As long as the contrast between the label and the background colour of the section satisfies the [contrast requirements for text](/knowledge/colours-and-contrast/text-contrast) this might be the most straight-forward solution.
+
+![The same pie chart, without patterns but with the labels placed inside the three sections](_media/pieChart-2.2.png)
+
```
I would also love to have an example with lines drawn from the chart elements to free floating labels outside, just as another typical inspiration.

AndreasMoesch replied:
As discussed, rather not “overload” the guide with more examples than necessary.

jmuheim commented on pages/setup/helper-tools/colour-contrast-analyser/README.md:
```diff
@@ -18,8 +18,8 @@ As it does not have an installer, you can simply extract it to a folder of your
 
-Note: if you choose to use the colour pickers instead, be sure to avoid anti-aliased colours, as this would have an impact on the result.
\ No newline at end of file
+For more information about colour contrast in general, there is a [dedicated chapter: Colors and contrast](/knowledge/colors-and-contrast/).
```
Forward link: `If you are really curious and want to learn more about this, skip ahead and read [Colors and contrast](/knowledge/colors-and-contrast/).`

See #237.

jmuheim commented on pages/knowledge/colours-and-contrast/text/README.md:
```diff
@@ -0,0 +1,79 @@
+- **Decorative text:** this could be the background pattern of an image that is created from random words.
+- **Incidental text:** this could be a random street sign somewhere in the background of a photograph.
+- **Disabled text:** this could be the label of a disabled element in an online form.
+
```
We should at least mention `placeholder` values here. They are a really nasty edge case, and they are talked about a lot already in other places (like https://www.nngroup.com/articles/form-design-placeholders/).

AndreasMoesch replied:
Standpoint A4A
* Mustn’t replace the label
* *if* it does it has to be like in Material Design (leaving the field on focus)
* For an *actual* placeholder too much contrast can confuse people (did I enter this?)
* Could also be in “forms” section.

**To Do:** Check https://www.nngroup.com/articles/form-design-placeholders/ and mention placeholder as they describe it.

AndreasMoesch replied:
I checked the WCAG page [Understanging Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html). There they state explicitly “The minimum contrast success criterion (1.4.3) applies to text in the page, including placeholder text and text that is shown when […]”

I added a sub title and short paragraph to the “Exceptions” section. I state what the specification says while also mentioning the general usability issues. For more information I link to the NNGroup article.

> What about placeholder text?
> 
> The WCAG's contrast requirement for text also applies to placeholder text. But it's important to mention that placeholders offers a bunch of usability and accessibility challenges beyond just adequate contrast – sometimes even because of adequate contrast. For more information see this [article about why placeholders in form fields are harmful](https://www.nngroup.com/articles/form-design-placeholders/)

jmuheim commented on pages/knowledge/colours-and-contrast/colour-is-not-enough/README.md:
```diff
@@ -0,0 +1,67 @@
+
+This problem can be solved by adding one or more visual alterations to the active element that have nothing to do with changing its colour. E.g. one could change the font weight or underline the active label.
+
+![navigation bar with one active item, visually differentiated by its blue text colour, bold font and underline](_media/navigation-currentItem-2.1.png)
```
So far, the file names always reflected the alternative texts as closely as possible (in a "sluggish" way):

```
![navigation bar with one active item, visually differentiated by its blue text colour, bold font and underline](_media/navigation-bar-with-one-active-item-visually-differentiated-by-its-blue-text-colour-bold-font-and-underline.png)
```

See #235.

AndreasMoesch replied:
**To Do:** Remove CamelCase. Improve file name, but don’t turn it into alt text.

gurix replied:
Pay attention with such long filenames. Recently I had the issue on a windows 10 that the filename was just cropped after downloading a PDF without any warnings. Therefore the suffix was gone and windows could not longer determine the file type. As result it just didn't open the file anymore.

AndreasMoesch replied:
Done. 
The file names are now longer than before but not full sentences.

AndreasMoesch committed:
Adds first set of content correcitons

Changes done on Hackday #12 together with Josh

AndreasMoesch committed:
Adds rest of content corrections

Closing the remaining review points from @jmuheim

AndreasMoesch committed:
Deletes unused pictures

AndreasMoesch commented:
@jmuheim I have taken care of all the open points from the review. Assigning to you for next steps.

jmuheim committed:
Update README.md

Typo

jmuheim committed:
Update README.md

jmuheim committed:
Update README.md

AndreasMoesch committed:
Resolves merge conflicts with master

AndreasMoesch committed:
Merge branch 'master' into content/colors-and-contrast
